File: Problems/No.1590_Make_Sum_Divisible_by_P/make_sum_divisible_by_p.py
class Solution(object):
    def minSubarray(self, nums, p):
        """
        :type nums: List[int]
        :type p: int
        :rtype: int
        """

        # parameters
        subarray_size = len(nums)
        remainder = 0
        preSum_index = {}

        # start computing
        # check the remainder (need to remove)
        remainder = sum(nums) % p
        if (remainder == 0):
            return 0
        else:
            pre_sum, find_gap = 0, 0
            preSum_index[0] = -1
            for i, num in enumerate(nums):
                # calculate prefix sum with %p & save the index
                pre_sum = (pre_sum + num) % p
                preSum_index[pre_sum] = i

                # find if the gap between the current prefix sum and the target remainder exist
                find_gap = (pre_sum - remainder + p) % p
                if find_gap in preSum_index:
                    subarray_size = min(subarray_size, i-preSum_index[find_gap])

            return subarray_size if (subarray_size < len(nums)) else -1

File: Problems/No.1590_Make_Sum_Divisible_by_P/make_sum_divisible_by_p.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        
        // parameters
        int subarray_size = nums.size();
        int remainder = 0; 
        std::unordered_map<int, int> preSum_index;

        // start computing
        // check the remainder (need to remove)
        for (auto &num: nums) remainder = (remainder + num) % p;
        if (remainder == 0) return 0;
        else {

            int pre_sum = 0, find_gap;
            preSum_index[0] = -1;
            for (int i=0; i<nums.size(); i++) {

                pre_sum = (pre_sum + nums[i]) % p;
                preSum_index[pre_sum] = i;

                find_gap = (pre_sum - remainder + p) % p;
                if (preSum_index.find(find_gap) != preSum_index.end())
                    subarray_size = std::min(subarray_size, i-preSum_index[find_gap]);

            }

            return (subarray_size < nums.size()) ? subarray_size : -1;

        }

    }
};
